### PR TITLE
keybase_daemon_rpc: keep the conn alive with CurrentSession calls

### PR DESCRIPTION
**Pull request by strib**
OnConnect is called lazily when a reconnection is triggered by an
outgoing RPC, and we need to send HelloIAm calls as soon as we can
after the connection is re-established.  So instead of waiting for
KBFS to make a call, have one going proactively in the background at
all times.

Issue: KBFS-1671

### CODE REVIEW

File: libkbfs/keybase_daemon_rpc.go [hunk 1]
```diff
@@ -308,9 +315,35 @@ func (k *KeybaseDaemonRPC) ShouldRetryOnConnect(err error) bool {
...
+	// connection comes back.
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			const sessionID = 0
+			_, err := k.sessionClient.CurrentSession(ctx, sessionID)
```
Comment (strib) [inline libkbfs/keybase_daemon_rpc.go]: @maxtaco @patrickxb: Is `CurrentSession` cheap enough to use as a keep-alive ping?  If not, is there a better choice?

Comment (patrickxb) [inline libkbfs/keybase_daemon_rpc.go]: It creates a service-wide lock, so not the greatest.

Can we just add a ping() you can call?

File: libkbfs/keybase_daemon_rpc.go [hunk 2]
```diff
@@ -308,9 +315,35 @@ func (k *KeybaseDaemonRPC) ShouldRetryOnConnect(err error) bool {
...
+			if err != nil && err.Error() != NoCurrentSessionExpectedError {
+				// Only log the error if it's not "no current session".
+				k.log.CDebugf(
```
Comment (akalin-keybase) [inline libkbfs/keybase_daemon_rpc.go]: Warning-worthy?

File: libkbfs/keybase_daemon_rpc.go [hunk 3]
```diff
@@ -308,9 +315,35 @@ func (k *KeybaseDaemonRPC) ShouldRetryOnConnect(err error) bool {
 	return !inputCanceled
 }
 
+func (k *KeybaseDaemonRPC) keepAliveLoop(ctx context.Context) {
+	// We need to send another HelloIAm message as soon as the
```
Comment (akalin-keybase) [inline libkbfs/keybase_daemon_rpc.go]: This comment is a bit confusing. The main problem is that we can't detect if a connection is dropped until the next RPC, right? Maybe word it something like:

```
If the connection is dropped, we need to re-connect and send another HelloIAm
 message. However, we can't actually detect when the connection is closed until
 KBFS makes another outgoing RPC, ...
```

File: libkbfs/keybase_daemon_rpc.go [hunk 4]
```diff
@@ -51,6 +53,11 @@ func NewKeybaseDaemonRPC(config Config, kbCtx Context, log logger.Logger, debug
 	conn := NewSharedKeybaseConnection(kbCtx, config, k)
 	k.fillClients(conn.GetClient())
 	k.shutdownFn = conn.Shutdown
+
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (akalin-keybase) [inline libkbfs/keybase_daemon_rpc.go]: May add a comment below in `newKeybaseDaemonRPCWithClient` that we don't need to bother doing a keepalive loop in that case.

File: libkbfs/keybase_daemon_rpc.go [hunk 5]
```diff
@@ -308,9 +315,35 @@ func (k *KeybaseDaemonRPC) ShouldRetryOnConnect(err error) bool {
...
+		}
+	}
+}
+
 // Shutdown implements the KeybaseService interface for KeybaseDaemonRPC.
 func (k *KeybaseDaemonRPC) Shutdown() {
 	if k.shutdownFn != nil {
 		k.shutdownFn()
 	}
+	k.keepAliveCancel()
```
Comment (akalin-keybase) [inline libkbfs/keybase_daemon_rpc.go]: Don't you need a nil check here? Or maybe fill it in with a do-nothing function in the test constructor?

File: libkbfs/keybase_daemon_rpc.go (hunk 1 above)
Comment (patrickxb) [inline libkbfs/keybase_daemon_rpc.go]: This one is cheap:

      OutOfDateInfo checkAPIServerOutOfDateWarning();

in config.avdl.  All it does is return data already in a struct in memory.

Kind of weird, though.  Obviously we could add ping() with little effort.

Comment (strib) [inline libkbfs/keybase_daemon_rpc.go]: Thanks @patrickxb, I'm happy to add an explicit `ping`.  Which protocol/file should I put it on the client side?

**Pull request merged**